File: src/main/java/leetcode/Problem1854.java
package leetcode;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * https://leetcode.com/problems/maximum-population-year/
 */
public class Problem1854 {
    public int maximumPopulation(int[][] logs) {
        Arrays.sort(logs, (a, b) -> {
            int cmp = Integer.compare(a[0], b[0]);
            if (cmp == 0) {
                return Integer.compare(a[1], b[1]);
            }
            return cmp;
        });
        Map</* year */ Integer, /* count */ Integer> map = new HashMap<>();
        for (int i = 0; i < logs.length; i++) {
            int birth1 = logs[i][0];
            if (map.containsKey(birth1)) {
                map.put(birth1, map.getOrDefault(birth1, 0) + 1);
                continue;
            }
            map.put(birth1, 1);
            for (int j = i - 1; j >= 0; j--) {
                int death2 = logs[j][1];
                if (birth1 < death2) {
                    map.put(birth1, map.getOrDefault(birth1, 0) + 1);
                }
            }
        }
        return Collections.max(map.entrySet(), (a, b) -> {
            int cmp = Integer.compare(a.getValue(),b.getValue()); // by count
            if (cmp == 0) {
                return Integer.compare(b.getKey(), a.getKey()); // by year
            }
            return cmp;
        }).getKey();
    }
}

File: src/main/rust/src/bin/problem1854.rs
use std::collections::HashMap;

// https://leetcode.com/problems/maximum-population-year/description/
pub fn maximum_population(logs: Vec<Vec<i32>>) -> i32 {
    let mut map: HashMap<i32, i32> = HashMap::new();
    for log in logs.iter() {
        let birth = log[0];
        let death = log[1];
        *map.entry(birth).or_insert(0) += 1;
        *map.entry(death).or_insert(0) -= 1;
    }
    let mut v: Vec<(i32, i32)> = map.into_iter().collect();
    v.sort();
    let mut answer = 0;
    let mut max = 0;
    let mut prefix_sum = 0;
    for i in 0..v.len() {
        if i == 0 {
            answer = v[i].0;
            prefix_sum = v[i].1;
            max = prefix_sum;
        } else {
            prefix_sum += v[i].1;
        }
        if max < prefix_sum {
            answer = v[i].0;
            max = prefix_sum;
        }
    }
    answer
}

fn main() {
    println!(
        "{}",
        maximum_population(vec![vec![1993, 1999], vec![2000, 2010]])
    ); // 1993
    println!(
        "{}",
        maximum_population(vec![vec![1950, 1961], vec![1960, 1971], vec![1970, 1981]])
    ); // 1960
}
